src/protocol/sort/apply.rs

```diff
@@ -11,77 +12,121 @@ use permutation::Permutation;
 
 #[embed_doc_image("apply", "images/sort/apply.png")]
 #[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+#[embed_doc_image("apply", "images/sort/apply.png")]
+#[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+
 /// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
 /// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
 /// ![Apply steps][apply]
-pub fn apply<T: Copy + Default, S>(mut permutation: Permutation, values: &mut S)
-where
-    S: AsMut<[T]>,
-{
-    permutation.apply_slice_in_place(values);
+pub fn apply<T: Copy + Default>(permutation: &[usize], values: &mut [T]) {
+    let mut permuted = bitvec![0; permutation.len()];
+    let mut tmp: T = T::default();
+
+    for i in 0..permutation.len() {
+        if permuted[i] == false {
+            mem::swap(&mut tmp, &mut values[i]);
+            let mut pos_i = i;
+            let mut pos_j = permutation[pos_i];
+            while pos_j != i {
+                values[pos_i] = values[pos_j];
+                pos_i = pos_j;
+                pos_j = permutation[pos_i];
+                permuted.set(pos_i, true);
+            }
+            mem::swap(&mut values[pos_i], &mut tmp);
+            permuted.set(i, true);
+        }
+    }
 }
 
 /// To compute `apply_inv` on values, permutation(i) can be regarded as the destination of i, i.e., the i-th item
 /// is moved by `apply_inv` to be the σ(i)-th item. Therefore, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is
 /// reordered into (D, C, A, B).
 /// ![Apply inv steps][apply_inv]
-pub fn apply_inv<T: Copy + Default, S>(mut permutation: Permutation, values: &mut S)
-where
-    S: AsMut<[T]>,
-{
-    permutation.apply_inv_slice_in_place(values);
+pub fn apply_inv<T: Copy + Default>(permutation: &[usize], values: &mut [T]) {
+    let mut permuted = bitvec![0; permutation.len()];
```

Comment:
After we push this, would be interesting to experiment with either passing this bit vector as an optional argument. We could re-use the same allocation for sequentially applied permutations and pass `None` if permutation is applied in parallel. I can create an issue if you think it is worth doing (I don't really remember how often we need  to apply permutations in parallel)

Reply:
to be more precise:

```rust
pub fn apply_inv<T: Copy + Default>(permutation: &[usize], values: &mut [T], buf: Option<&mut BitSlice>) {
   // if buf is set, re-use it for to track permutations
   // if it is not set, allocate and proceed
}

Reply:
yup. This is one of the reasons I went with this refactor. I suspect we can re-use one buffer for this. As you say - next diff.